File: 10. Regular Expression Matching/10.py
class Solution:

    def isMatch(self, s: str, p: str) -> bool:
        self.memo = {}
        return self.dp(s, 0, p, 0)

    def dp(self, s: str, i: int, p: str, j: int) -> bool:
        m, n = len(s), len(p)
        if j == n:
            return i == m
        if i == m:
            if (n - j) % 2 == 1:
                return False
            while j < n - 1:
                if p[j + 1] != '*':
                    return False
                j += 2
            return True
        key = str(i) + str(j)
        if key in self.memo:
            return self.memo[key]
        self.ans = False
        if p[j] == '.' or p[j] == s[i]:
            if j < n - 1 and p[j + 1] == '*':
                self.ans = self.dp(s, i + 1, p, j) or self.dp(s, i, p, j + 2)
            else:
                self.ans = self.dp(s, i + 1, p, j + 1)
        else:
            if j < n - 1 and p[j + 1] == '*':
                self.ans = self.dp(s, i, p, j + 2)
            else:
                return False
        self.memo[key] = self.ans
        return self.ans

File: 10. Regular Expression Matching/10.kt
class Solution {
    fun isMatch(s: String, p: String): Boolean {
        val memo = mutableMapOf<String, Boolean>()
        fun dp(i: Int, j: Int): Boolean {
            var k = j
            val m = s.length
            val n = p.length
            if (k == n) return i == m
            if (i == m) {
                if (((n - k) % 2) == 1) {
                    return false
                }
                while (k < n - 1) {
                    if (p[k + 1] != '*') {
                        return false
                    }
                    k += 2
                }
                return true
            }
            val key="$i$k"
            if (memo.containsKey(key)) {
                return memo[key]!!
            }
            var ans = false
            ans = if (p[k] == '.' || p[k] == s[i]) {
                if (j + 1 < n && p[k + 1] == '*') {
                    dp(i + 1, k) || dp(i, k + 2);
                } else {
                    dp(i + 1, k + 1);
                }
            } else {
                if (j + 1 < n && p[k + 1] == '*') {
                    dp(i, k + 2);
                } else {
                    false;
                }
            }
            memo[key] = ans;
            return ans;
        }
        return dp(0, 0)
    }
}